cpp/src/parquet/encoding_test.cc

```diff
@@ -1955,5 +1955,285 @@ TEST(DeltaLengthByteArrayEncodingAdHoc, ArrowDirectPut) {
+    std::vector<int32_t> arrays;
+    auto data_ptr = checked_cast<::arrow::Int32Array*>(lengths.get());
+    for (int i = 0; i < lengths->length(); ++i) {
+      arrays.push_back(data_ptr->GetView(i));
+    }
+    return arrays;
+  };
+
+  auto CheckDecode = [](std::shared_ptr<Buffer> buf,
+                        std::shared_ptr<::arrow::Array> values) {
+    int num_values = static_cast<int>(values->length());
+    auto decoder = MakeTypedDecoder<ByteArrayType>(Encoding::DELTA_BYTE_ARRAY);
+    decoder->SetData(num_values, buf->data(), static_cast<int>(buf->size()));
+
+    typename EncodingTraits<ByteArrayType>::Accumulator acc;
+    if (::arrow::is_string(values->type()->id())) {
+      acc.builder = std::make_unique<::arrow::StringBuilder>();
+    } else {
+      acc.builder = std::make_unique<::arrow::BinaryBuilder>();
+    }
+
+    ASSERT_EQ(num_values,
+              decoder->DecodeArrow(static_cast<int>(values->length()),
+                                   static_cast<int>(values->null_count()),
+                                   values->null_bitmap_data(), values->offset(), &acc));
+
+    std::shared_ptr<::arrow::Array> result;
+    ASSERT_OK(acc.builder->Finish(&result));
+    ASSERT_EQ(num_values, result->length());
+    ASSERT_OK(result->ValidateFull());
+
+    auto upcast_result = CastBinaryTypesHelper(result, values->type());
+    ::arrow::AssertArraysEqual(*values, *upcast_result);
+  };
+
+  auto checkEncodeDecode = [&](std::string_view values,
+                               std::shared_ptr<::arrow::Array> prefix_lengths,
+                               std::shared_ptr<::arrow::Array> suffix_lengths,
+                               std::string_view suffix_data) {
+    CheckEncode(::arrow::ArrayFromJSON(::arrow::utf8(), values), prefix_lengths,
+                suffix_lengths, suffix_data);
+    CheckEncode(::arrow::ArrayFromJSON(::arrow::large_utf8(), values), prefix_lengths,
+                suffix_lengths, suffix_data);
+    CheckEncode(::arrow::ArrayFromJSON(::arrow::binary(), values), prefix_lengths,
+                suffix_lengths, suffix_data);
+    CheckEncode(::arrow::ArrayFromJSON(::arrow::large_binary(), values), prefix_lengths,
+                suffix_lengths, suffix_data);
+
+    auto encoded = ::arrow::ConcatenateBuffers({DeltaEncode(arrayToI32(prefix_lengths)),
+                                                DeltaEncode(arrayToI32(suffix_lengths)),
+                                                std::make_shared<Buffer>(suffix_data)})
+                       .ValueOrDie();
+
+    CheckDecode(encoded, ::arrow::ArrayFromJSON(::arrow::utf8(), values));
+    CheckDecode(encoded, ::arrow::ArrayFromJSON(::arrow::large_utf8(), values));
+    CheckDecode(encoded, ::arrow::ArrayFromJSON(::arrow::binary(), values));
+    CheckDecode(encoded, ::arrow::ArrayFromJSON(::arrow::large_binary(), values));
+  };
+
+  {
+    auto values = R"(["axis", "axle", "babble", "babyhood"])";
```

Comment:
Should we add a test for non-ASCII but UTF8-conforming characters?

Reply:
I'll push an example soon.

Reply:
I'm trying this example but encoded result currently doesn't match up:
```cpp
  {
    auto values = R"(["καλημέρα", "καμηλιέρη", "καμηλιέρη", "καλημέρα"])";
    auto prefix_lengths = ::arrow::ArrayFromJSON(::arrow::int32(), R"([0, 2, 9, 2])");
    auto suffix_lengths = ::arrow::ArrayFromJSON(::arrow::int32(), R"([8, 7, 0, 6])");
    constexpr std::string_view suffix_data = "καλημέραμηλιέρηλημέρα";
    CheckEncodeDecode(values, prefix_lengths, suffix_lengths, suffix_data);
  }
```

```
Value of: prefix_lengths_buf->Equals(*encoded_prefix_length_buf)
  Actual: false
Expected: true

Value of: suffix_lengths_buf->Equals(*encoded_suffix_length_buf)
  Actual: false
Expected: true

Expected equality of these values:
  value
    Which is: "\xCE\xBA\xCE\xB1\xCE\xBB\xCE\xB7\xCE\xBC\xCE\xAD\xCF\x81\xCE\xB1\xCE\xBC\xCE\xB7\xCE\xBB\xCE\xB9\xCE\xAD\xCF\x81\xCE\xB7\xCE\xBB\xCE\xB7\xCE\xBC\xCE\xAD\xCF\x81\xCE\xB1"
    As Text: "καλημέραμηλιέρηλημέρα"
  encoded_values_buf->ToString()
    Which is: "\xCE\xBA\xCE\xB1\xCE\xBB\xCE\xB7\xCE\xBC\xCE\xAD\xCF\x81\xCE\xB1\xBC\xCE\xB7\xCE\xBB\xCE\xB9\xCE\xAD\xCF\x81\xCE\xB7\xBB\xCE\xB7\xCE\xBC\xCE\xAD\xCF\x81\xCE\xB1"
```

Reply:
Ah, actually this passes, so I think we're good:
```
  {
    auto values = R"(["καλημέρα", "καμηλιέρη", "καμηλιέρη", "καλημέρα"])";
    auto prefix_lengths = ::arrow::ArrayFromJSON(::arrow::int32(), R"([0, 5, 18, 5])");
    auto suffix_lengths = ::arrow::ArrayFromJSON(::arrow::int32(), R"([16, 13, 0, 11])");
    const std::string suffix_data = "καλημέρα\xbcηλιέρη\xbbημέρα";
    CheckEncodeDecode(values, prefix_lengths, suffix_lengths, suffix_data);
  }
```